server/src/fstar_connection.ts

```diff
@@ -0,0 +1,512 @@
+	close() {
+		this.fstar.proc.kill();
+	}
+
+	async restartSolver(debug: boolean) {
+		this.fstar.killZ3SubProcess(debug);
+
+		// Wait for a second for processes to die before restarting the solver
+		await setTimeout(1000);
+		this.restartSolverRequest(debug);
+	}
+
+	// Gets the configuration for the F* process.
+	fstar_config() : FStarConfig {
+		return this.fstar.config;
+	}
+
+	////////////////////////////////////////////////////////////////////////////////////
+	// Methods to send requests to F*
+	////////////////////////////////////////////////////////////////////////////////////
+
+	// Sends a request to the F* process.
+	//
+	// @param expect_response indicates if a response from F* is expected for
+	// this request.
+	private async request(msg: any, debug: boolean, expect_response = true, is_stream = false): Promise<any> {
+		const qid = this.last_query_id + 1;
+		this.last_query_id = qid;
+		msg["query-id"] = '' + qid;
+		const text = JSON.stringify(msg);
+		if (debug) {
+			console.log(">>> " + text);
+		}
+		if (this.fstar.proc.exitCode != null) {
+			const process_name = this.fstar.lax ? "flycheck" : "checker";
+			const error_msg = "ERROR: F* " + process_name + " process exited with code " + this.fstar.proc.exitCode;
+			// TODO(klinvill): should we throw exceptions here or reject the
+			// returned promise? Although I believe the semantics of async
+			// functions are such that this will automatically result in a
+			// rejected promise?
+			throw new Error(error_msg);
+		}
+
+		try {
+			this.fstar.proc?.stdin?.write(text);
+			this.fstar.proc?.stdin?.write("\n");
+		} catch (e) {
+			// TODO(klinvill): should we throw exceptions here or reject the
+			// returned promise? Although I believe the semantics of async
+			// functions are such that this will automatically result in a
+			// rejected promise?
+			const msg = "ERROR: Error writing to F* process: " + e;
+			throw new Error(msg);
+		}
+
+		const internal_promise =  withResolvers();
+		if (expect_response) {
+			// Only keep track of responses if we expect a response
+			this.pending_responses.set(qid, {ip: internal_promise, is_stream});
+		}
+		// TODO(klinvill): Should we always return a promise even if no response
+		// is expected? In that case the promise is just ignored.
+		return internal_promise.promise;
+	}
+
+	// Send a request to F* to check the given code.
+	async fullBufferRequest(code: string, kind: 'full' | 'lax' | 'cache' | 'reload-deps', withSymbols: boolean, debug: boolean): partialResult<IdeProgress> {
+		if (!this.fstar.supportsFullBuffer) {
+			throw new Error("ERROR: F* process does not support full-buffer queries");
+		}
+		const query: FullBufferQuery = {
+			query: "full-buffer",
+			args: {
+				kind,
+				"with-symbols": withSymbols,
+				code: code,
+				line: 0,
+				column: 0
+			}
+		};
+
+		const expect_response = true;
+		const is_stream = true;
+		return this.request(query, debug, expect_response, is_stream);
+	}
```

Comment:
I'd factor out the common parts of the two functions here:

```suggestion
        async fullBufferQuery(query: FullBufferQuery): partialResult<IdeProgress> {
		if (!this.fstar.supportsFullBuffer) {
			throw new Error("ERROR: F* process does not support full-buffer queries");
		}
		const expect_response = true;
		const is_stream = true;
		return this.request(query, expect_response, is_stream);
        }

	async fullBufferRequest(code: string, kind: 'full' | 'lax' | 'cache' | 'reload-deps', withSymbols: boolean): partialResult<IdeProgress> {
		return this.fullBufferQuery({
			query: "full-buffer",
			args: {
				kind,
				"with-symbols": withSymbols,
				code: code,
				line: 0,
				column: 0,
			},
		});
	}
```